p2p: Make timeout mockable and type safe, speed up test

maflcko opened the pull request:
Use type-safe time for better code readability/maintainability and mockable time for better testability. This speeds up the p2p_timeout test.

This is also a bugfix for intermittent test issues like: https://cirrus-ci.com/task/4769904156999680?command=ci#L2836

Fixes #20654

Commit:
p2p: Make timeout mockable and type safe, speed up test

practicalswift commented:
Concept ACK: mockable is good

luke-jr commented:
Doesn't this defeat the purpose of the test?

jnewbery commented:
Concept ACK. There's a lot of refactor/tidyup in this PR. I wonder if that can be split out to make this more focused on the functional changes.

Commit:
scripted-diff: Rename m_last_send and m_last_recv

-BEGIN VERIFY SCRIPT-

ren() { sed -i "s/\<$1\>/$2/g" $(git grep -l "$1" ./src) ; }

ren nLastSend m_last_send
ren nLastRecv m_last_recv

-END VERIFY SCRIPT-

MarcoFalke commented:
Rebased (for fun only)

mzumsande reviewed (commented):
Concept ACK

Reviewed the non-GUI-related parts so far and played with `p2p_timeouts.py` - looks good to me, just some nits.

MarcoFalke commented on src/net.cpp:
```diff
@@ -1317,31 +1317,33 @@ void CConnman::NotifyNumConnectionsChanged()
 
-    if (node.m_last_recv == 0 || node.m_last_send == 0) {
-        LogPrint(BCLog::NET, "socket no message in first %i seconds, %d %d peer=%d\n", m_peer_connect_timeout, node.m_last_recv != 0, node.m_last_send != 0, node.GetId());
+    if (last_recv.count() == 0 || last_send.count() == 0) {
```
`last_recv` is a bit like `std::optional`. I use `.count()`, when `.has_value()` would be used and I use `count_seconds()`, when `.value()` would be used.

MarcoFalke commented on src/net.cpp:
```diff
@@ -1317,31 +1317,33 @@ void CConnman::NotifyNumConnectionsChanged()
-    // Use non-mockable system time (otherwise these timers will pop when we
-    // use setmocktime in the tests).
-    int64_t now = GetTimeSeconds();
+    // Test that see disconnects after using mocktime can start with a large
```
thx, fixed

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -4312,9 +4312,10 @@ void PeerManagerImpl::CheckForStaleTipAndEvictPeers()
+    if (m_connman.ShouldRunInactivityChecks(node_to, std::chrono::duration_cast<std::chrono::seconds>(now)) &&
         peer.m_ping_nonce_sent &&
-        now > peer.m_ping_start.load() + std::chrono::seconds{TIMEOUT_INTERVAL}) {
+        now > peer.m_ping_start.load() + std::chrono::seconds{TIMEOUT_INTERVAL})
```
Nice, fixed.

laanwj commented on src/net_processing.cpp:
```diff
@@ -4312,9 +4312,10 @@ void PeerManagerImpl::CheckForStaleTipAndEvictPeers()
         peer.m_ping_nonce_sent &&
-        now > peer.m_ping_start.load() + std::chrono::seconds{TIMEOUT_INTERVAL}) {
+        now > peer.m_ping_start.load() + TIMEOUT_INTERVAL)
+    {
```
More consistent to keep the `{` on the same line?

MarcoFalke replied:
I think (and other seem to agree, see https://github.com/bitcoin/bitcoin/issues/21735) that multiline if conditions are hard to read because there is no break between condition and body, so changed it here.

Happy to revert, though.

laanwj replied:
I don't particularly want to get into discussion about code style, it just seemed inconsistent to me (as all other `if` cases have the `{` on the same line), if you have a good reason for it it's fine with me.

laanwj commented:
Code review ACK fadc0c80ae4e526fb2b503f7cc02f6122aaf1de5

naumenkogs commented:
ACK fadc0c80ae4e526fb2b503f7cc02f6122aaf1de5
